File: src/main/scala/mmlTests/a201901a/java/gpt4/GraphFactoryImpl.java
package mmlTests.a201901a.java.gpt4;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import mmlTests.a201901a.java.*;
import mmlTests.shared.java.Pair;

public class GraphFactoryImpl implements GraphFactory {

    @Override
    public <X> Graph<X> createDirectedChain(List<X> nodes) {
        return new GraphImpl<>(new HashSet<>(nodes), getEdges(nodes, false));
    }

    @Override
    public <X> Graph<X> createBidirectionalChain(List<X> nodes) {
        return new GraphImpl<>(new HashSet<>(nodes), getEdges(nodes, true));
    }

    @Override
    public <X> Graph<X> createDirectedCircle(List<X> nodes) {
        List<Pair<X,X>> edges = getEdges(nodes, false);
        edges.add(new Pair<>(nodes.get(nodes.size()-1), nodes.get(0)));
        return new GraphImpl<>(new HashSet<>(nodes), edges);
    }

    @Override
    public <X> Graph<X> createBidirectionalCircle(List<X> nodes) {
        List<Pair<X,X>> edges = getEdges(nodes, true);
        edges.add(new Pair<>(nodes.get(nodes.size()-1), nodes.get(0)));
        return new GraphImpl<>(new HashSet<>(nodes), edges);
    }

    @Override
    public <X> Graph<X> createDirectedStar(X center, Set<X> nodes) {
        List<Pair<X,X>> edges = nodes.stream().map(n -> new Pair<>(center, n)).collect(Collectors.toList());
        nodes.add(center);
        return new GraphImpl<>(nodes, edges);
    }

    @Override
    public <X> Graph<X> createBidirectionalStar(X center, Set<X> nodes) {
        List<Pair<X,X>> edges = nodes.stream().flatMap(n -> Stream.of(new Pair<>(n, center), new Pair<>(center, n))).collect(Collectors.toList());
        nodes.add(center);
        return new GraphImpl<>(nodes, edges);
    }

    @Override
    public <X> Graph<X> createFull(Set<X> nodes) {
        List<Pair<X,X>> edges = new ArrayList<>();
        for(X node1 : nodes) {
            for(X node2 : nodes) {
                if(!node1.equals(node2)) {
                    edges.add(new Pair<>(node1, node2));
                }
            }
        }
        return new GraphImpl<>(nodes, edges);
    }

    @Override
    public <X> Graph<X> combine(Graph<X> g1, Graph<X> g2) {
        Set<X> combinedNodes = new HashSet<>(g1.getNodes());
        combinedNodes.addAll(g2.getNodes());

        List<Pair<X,X>> combinedEdges = new ArrayList<>(g1.getEdgesStream().collect(Collectors.toList()));
        combinedEdges.addAll(g2.getEdgesStream().collect(Collectors.toList()));

        return new GraphImpl<>(combinedNodes,combinedEdges);
    }

    private <X> List<Pair<X,X>> getEdges(List<X> nodes, boolean bidirectional) {
        List<Pair<X,X>> edges = new ArrayList<>();
        for(int i=0; i<nodes.size()-1; i++) {
            edges.add(new Pair<>(nodes.get(i), nodes.get(i+1)));
            if(bidirectional) {
                edges.add(new Pair<>(nodes.get(i+1), nodes.get(i)));
            }
        }
        return edges;
    }

    private class GraphImpl<X> implements Graph<X> {
        private final Set<X> nodes;
        private final List<Pair<X, X>> edges;

        GraphImpl(Set<X> nodes, List<Pair<X, X>> edges) {
            this.nodes = nodes;
            this.edges = edges;
        }

        @Override
        public Set<X> getNodes() {
            return nodes;
        }

        @Override
        public boolean edgePresent(X start, X end) {
            return edges.contains(new Pair<>(start, end));
        }

        @Override
        public int getEdgesCount() {
            return edges.size();
        }

        @Override
        public Stream<Pair<X, X>> getEdgesStream() {
            return edges.stream();
        }
    }
}

File: src/main/scala/mmlTests/a201901a/scala/gpt4/GraphFactoryImpl.scala
package mmlTests.a201901a.scala.gpt4

import mmlTests.a201901a.scala.{Graph, GraphFactory}
import mmlTests.a201901a.scala.gpt4.GraphImpl

import java.util
import java.util.stream.Stream
import mmlTests.shared.scala.Pair

class GraphFactoryImpl() extends GraphFactory {

  private def createChain[X](nodes: util.List[X], bidirectional: Boolean): Graph[X] = {
    val g = new GraphImpl[X]()
    val it = nodes.iterator()
    if (it.hasNext) {
      var previous = it.next()
      g.nodes.add(previous)

      while (it.hasNext) {
        val current = it.next()
        g.nodes.add(current)
        g.edges.putIfAbsent(previous, new util.HashSet[X]())
        g.edges.get(previous).add(current)

        if (bidirectional) {
          g.edges.putIfAbsent(current, new util.HashSet[X]())
          g.edges.get(current).add(previous)
        }

        previous = current
      }
    }
    g
  }

  private def createCircle[X](nodes: util.List[X], bidirectional: Boolean): Graph[X] = {
    if (nodes.isEmpty) return new GraphImpl[X]()
    val g = createChain(nodes, bidirectional)
    val first = nodes.get(0)
    val last = nodes.get(nodes.size() - 1)

//    g.edges.putIfAbsent(last, new util.HashSet[X])
//    g.edges.get(last).add(first)
//    if (bidirectional) {
//      g.edges.putIfAbsent(first, new util.HashSet[X])
//      g.edges.get(first).add(last)
//    }

    g
  }

  private def createStar[X](center: X, nodes: util.Set[X], bidirectional: Boolean): Graph[X] = {
    val g = new GraphImpl[X]()
    g.nodes.add(center)
    g.nodes.addAll(nodes)
    g.edges.putIfAbsent(center, new util.HashSet[X])
    nodes.forEach { node =>
      g.edges.get(center).add(node)
      if (bidirectional) {
        g.edges.putIfAbsent(node, new util.HashSet[X])
        g.edges.get(node).add(center)
      }
    }
    g
  }

  override def createDirectedChain[X](nodes: util.List[X]): Graph[X] = createChain(nodes, bidirectional = false)

  override def createBidirectionalChain[X](nodes: util.List[X]): Graph[X] = createChain(nodes, bidirectional = true)

  override def createDirectedCircle[X](nodes: util.List[X]): Graph[X] = createCircle(nodes, bidirectional = false)

  override def createBidirectionalCircle[X](nodes: util.List[X]): Graph[X] = createCircle(nodes, bidirectional = true)

  override def createDirectedStar[X](center: X, nodes: util.Set[X]): Graph[X] = createStar(center, nodes, bidirectional = false)

  override def createBidirectionalStar[X](center: X, nodes: util.Set[X]): Graph[X] = createStar(center, nodes, bidirectional = true)

  override def createFull[X](nodes: util.Set[X]): Graph[X] = {
    val g = new GraphImpl[X]()
    g.nodes.addAll(nodes)
    nodes.forEach { node1 =>
      g.edges.putIfAbsent(node1, new util.HashSet[X])
      nodes.forEach { node2 =>
        if (!node1.equals(node2)) g.edges.get(node1).add(node2)
      }
    }
    g
  }

  def combine[X](g1: Graph[X], g2: Graph[X]): Graph[X] = {
    val nodes = new util.HashSet[X]()
    nodes.addAll(g1.getNodes)
    nodes.addAll(g2.getNodes)

    val edges = new util.HashMap[X, util.HashSet[X]]
//    g1.edges.forEach((k, v) => edges.put(k, new util.HashSet[X](v)))
//    g2.edges.forEach((k, v) => edges.get(k).addAll(v))

    new GraphImpl[X](nodes, edges)
  }
}